python: add version() to get running version

davidhewitt opened the pull request:
This PR adds a new API `Python::version()` which is a safe convenience wrapper around `Py_GetVersion()`.

Now that we offer abi3 wheels which may need to support multiple different Python versions, such an API will come in helpful. We saw this already in #1320.

davidhewitt commented:
PyPy failure is upstream bug in `setup-python`, but we might not want to merge this until this is resolved as I'd like to know that this API works correctly with PyPy.

kngwyu reviewed (changes requested):
Thanks!

kngwyu commented on src/python.rs:
```diff
@@ -7,12 +7,129 @@ use crate::gil::{self, GILGuard, GILPool};
+#[derive(Debug)]
+pub struct PythonVersion {
+    major: u8,
+    minor: (u8, Option<&'static str>),
+    patch: Option<(u8, Option<&'static str>)>,
```
Also, I think we can make all members transparent (=`pub`) with this style.

davidhewitt replied:
I like this idea, let's try it. And also I agree it's probably correct to assume `patch == 0`.

It's possible all version strings we will see in practice include a patch, though the example for `Py_GetVersion` doesn't!

davidhewitt commented on src/python.rs:
```diff
@@ -618,4 +760,39 @@ mod test {
+        assert!(PythonVersion::from_str("3.5+") >= (3, 5, 0));
+        assert!(PythonVersion::from_str("3.5+") == (3, 5, 0));
+        assert!(PythonVersion::from_str("3.5+") != (3, 5, 1));
+        assert!(PythonVersion::from_str("3.5+") == (3, 5, 0));
```
Yes, good spot.

davidhewitt commented on src/python.rs:
```diff
@@ -7,12 +7,129 @@ use crate::gil::{self, GILGuard, GILPool};
+        self.patch.as_ref().map(|patch| patch.0)
+    }
+
+    /// Parses a hard-coded Python interpreter version string.
```
This is a private API so I don't think it needs a full doctest example snippet, I will add a note of the form as you suggest though.

davidhewitt commented:
Thanks for the review, I pushed some changes to address.

Also I renamed this API to `PythonVersionInfo` and `Python::version_info()`, and also added `Python::version() -> &'static str`. This matches better with `sys.version_info` and `sys.version`.

kngwyu commented on src/python.rs:
```diff
@@ -7,12 +7,112 @@ use crate::gil::{self, GILGuard, GILPool};
+
+impl PartialOrd<(u8, u8)> for PythonVersionInfo {
+    fn partial_cmp(&self, other: &(u8, u8)) -> Option<std::cmp::Ordering> {
+        match self.major.cmp(&other.0) {
```
I think you can compare the tuples direcly here, like
```
(self.major, self.minor).partial_cmp(other)
```

davidhewitt commented:
Thanks, simplified `partial_cmp` implementations and force-pushed.

alex commented on src/python.rs:
````diff
@@ -302,6 +386,49 @@ impl<'p> Python<'p> {
+    /// ```
+    pub fn version(self) -> &'static str {
+        unsafe {
+            CStr::from_ptr(ffi::Py_GetVersion() as *const c_char)
````
I'm not sure treating the result of `Py_GetVersion` as ` `&'static str` is safe. Because of the way it's implemented, if you have a thread calling `version()` at the same time as another thread is attempting to read the contents of the str, you have a data race: https://github.com/python/cpython/blob/master/Python/getversion.c

Is there a reason it wouldn't work to give it the `'py` lifetime?

davidhewitt replied:
Ah, yes the claim in the docs it was "static storage" misled me. Nice catch - I'll change the lifetime 👍

alex commented on src/python.rs:
```diff
@@ -7,12 +7,99 @@ use crate::gil::{self, GILGuard, GILPool};
+    pub major: u8,
+    pub minor: u8,
+    pub patch: u8,
+    pub suffix: Option<String>,
```
This could be `Option<&'a str>` if you want to avoid the malloc.

davidhewitt replied:
Yeah; I was trying to keep lifetimes simple on the type but perhaps it's worth doing.

kngwyu commented:
Thanks!
